pkg/cmd/checkendpoints/controller/controller.go

```diff
@@ -0,0 +1,255 @@
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/trace"
+)
+
+// PodNetworkConnectivityCheckController continuously performs network connectivity
+// checks and reports the results.
+type PodNetworkConnectivityCheckController interface {
+	factory.Controller
+}
+
+// controller implements a PodNetworkConnectivityCheckController that discovers the list of endpoints to
+// check by looking for PodNetworkConnectivityChecks in a given namespace, for a specific pod. Updates to
+// the PodNetworkConnectivityCheck status are queued up and handled asynchronously such that disruptions
+// to the ability to update the PodNetworkConnectivityCheck status do not disrupt the ability to perform
+// the connectivity checks.
+type controller struct {
+	factory.Controller
+	podName      string
+	podNamespace string
+	checksGetter operatorcontrolplaneclientv1alpha1.PodNetworkConnectivityCheckInterface
+	checkLister  v1alpha1.PodNetworkConnectivityCheckNamespaceLister
+	recorder     events.Recorder
+	// each PodNetworkConnectivityCheck gets its own PodNetworkConnectivityCheckStatusUpdater
+	updaters map[string]PodNetworkConnectivityCheckStatusUpdater
+}
+
+// Returns a new PodNetworkConnectivityCheckController that performs network connectivity checks
+// as specified in the PodNetworkConnectivityChecks defined in the specified namespace, for the specified pod.
+func New(podName, podNamespace string, checksGetter operatorcontrolplaneclientv1alpha1.PodNetworkConnectivityChecksGetter, checkInformer alpha1.PodNetworkConnectivityCheckInformer, recorder events.Recorder) PodNetworkConnectivityCheckController {
+	c := &controller{
+		podName:      podName,
+		podNamespace: podNamespace,
+		checksGetter: checksGetter.PodNetworkConnectivityChecks(podNamespace),
+		checkLister:  checkInformer.Lister().PodNetworkConnectivityChecks(podNamespace),
+		recorder:     recorder,
+		updaters:     map[string]PodNetworkConnectivityCheckStatusUpdater{},
+	}
+	c.Controller = factory.New().
+		WithSync(c.Sync).
+		WithInformers(checkInformer.Informer()).
+		ResyncEvery(1*time.Second).
+		ToController("check-endpoints", recorder)
+	return c
+}
+
+// Sync ensures that the status updaters for each PodNetworkConnectivityCheck is started
+// and then performs each check.
+func (c *controller) Sync(ctx context.Context, syncContext factory.SyncContext) error {
+	checkList, err := c.checkLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	// filter list of checks for current pod
+	var checks []*operatorcontrolplanev1alpha1.PodNetworkConnectivityCheck
+	for _, check := range checkList {
+		if check.Spec.SourcePod == c.podName {
+			checks = append(checks, check)
+		}
+	}
+
+	// create & start status updaters if needed
+	for _, check := range checks {
+		if updater := c.updaters[check.Name]; updater == nil {
+			c.updaters[check.Name] = NewPodNetworkConnectivityCheckStatusUpdater(c, check.Name, c.recorder)
+			c.updaters[check.Name].Start(ctx)
```

Comment:
if it's just starting these, why does it have a 1 second sync?

Reply:
> if it's just starting these, why does it have a 1 second sync?

Assuming it's because you're issuing updates below, it seems like code may simplify simply by keeping track of "results since last update" and waiting until it hits X where X is 20 or so.